### PR TITLE
[FN] Wallet conflicting transaction detection & indexing

### PR DESCRIPTION
**Pull request by zeptin**
Includes the test from #3593 with some bugfixes.

Do not merge yet, refactoring appears to have affected the validity of the test, as it passes but does not seem to hit the expected code path any more.

### CODE REVIEW

**Comment by dangershony** [comment]
Probably if introducing indexing to the wallet this should be in its own PR

The mechanism of detecting a double spend in the wallet also applies to mempool,
Right now mempool will reject a double spend however if the trx that is rejected from mempool is included in a block I dont think we clear the losing trx from mempool, we should be looking at inputs of all trx confirmed in a block (we might be doing this already so need to investigate).

**Comment by bokobza** [comment]
I don't think this is the proper way to do it.
In fact, I think we should do nothing at all.
It's a lot of changes and it will probably double the time it takes to
process transactions, only to fix something that can only be created on
purpose by someone, and that can be resolved by resyncing the wallet when
it happens.

We just need to make sure that when a transaction is created, the input is
marked as spent, so that no user can recreate this issue by mistake.

**Comment by dangershony** [comment]
You have to be carful with changes to the wallet @zeptin, Jeremy is watching 😄

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 1]
```diff
@@ -1003,6 +1010,42 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
...
+                foreach (TxIn input in transaction.Inputs)
+                {
+                    // See if this input is being used by another transaction.
+                    if (this.inputLookup.TryGetValue(input.PrevOut, out WalletIndexData indexData))
+                    {
+                        if (indexData.TransactionData.Id == hash)
+                            continue; // It's the same transaction, ignore.
+
+                        // Incoming transaction confirmed, existing transaction unconfirmed, delete existing
+                        if ((block != null) && (indexData.TransactionData.BlockHash == null))
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Perhaps rename `block` to `confirmedBlock`? So that we know that we are working with a transaction that is in a confirmed block?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 2]
```diff
@@ -1003,6 +1010,42 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
...
+                // Do a pre-scan of the transaction inputs to see if they're used in other transactions already
+                foreach (TxIn input in transaction.Inputs)
+                {
+                    // See if this input is being used by another transaction.
+                    if (this.inputLookup.TryGetValue(input.PrevOut, out WalletIndexData indexData))
+                    {
+                        if (indexData.TransactionData.Id == hash)
+                            continue; // It's the same transaction, ignore.
+
+                        // Incoming transaction confirmed, existing transaction unconfirmed, delete existing
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Perhaps just add to this comment that `indexData` is data/transactions that is in our wallet.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 3]
```diff
@@ -1003,6 +1010,42 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
...
+                        if ((block != null) && (indexData.TransactionData.BlockHash == null))
+                        {
+                            this.RemoveTransactionsByIdsLocked(indexData.Wallet.Name, new[] { indexData.TransactionData.Id });
+                        }
+
+                        // Incoming transaction unconfirmed, existing transaction unconfirmed, delete existing (so that only 1 is left - either could be the 'correct' version)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: So here we effectively take the "superseding" transaction? I.e. we are replacing the one in lookup dictionary?

File: src/Stratis.Bitcoin.Features.Wallet/WalletIndexData.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    public class WalletIndexData
+    {
+        public readonly Wallet Wallet;
+        public readonly HdAccount Account;
+        public readonly HdAddress Address;
+        public readonly TransactionData TransactionData;
+
+        public WalletIndexData(Wallet wallet, HdAccount account, HdAddress address, TransactionData transactionData)
+        {
+            this.Wallet = wallet;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletIndexData.cs]: Perhaps add Guards here?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 5]
```diff
@@ -1105,6 +1149,27 @@ private void AddTransactionToWallet(Transaction transaction, TxOut utxo, int? bl
...
+                // Workaround for now to populate index as the address lookup doesn't give us all the requisite data
+                // TODO: Either amend address lookup to provide additional fields or optimise in some other way
+                bool finish = false;
+                foreach (Wallet wallet in this.Wallets)
+                {
+                    if (finish) break;
+                    foreach (HdAccount account in wallet.GetAccounts())
+                    {
+                        if (finish) break;
+                        foreach (HdAddress address1 in account.GetCombinedAddresses())
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Rename `address1` to `accountAddress`?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 6]
```diff
@@ -1105,6 +1149,27 @@ private void AddTransactionToWallet(Transaction transaction, TxOut utxo, int? bl
...
+                        {
+                            if (finish) break;
+                            if (address == address1)
+                            {
+                                this.AddTxLookupLocked(wallet, account, address, newTransaction);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: So just to understand the logic... If teh address associated to the script matches an address in the wallet we add it?

Can't we do some fancy LINQ statement using `SelectMany` here:

`var applicableAddresses = this.Wallets.SelectMany(w=w.GetAccounts()).SelectMany(a=>a.GetCombinedAddresses())`

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 7]
```diff
@@ -1524,6 +1603,62 @@ public void RefreshInputKeysLookupLock()
...
+        /// Reloads the transactions we're tracking in memory for faster lookups.
+        /// </summary>
+        private void RefreshTxLookupLocked()
+        {
+            lock (this.lockObject)
+            {
+                this.txLookup = new Dictionary<uint256, WalletIndexData>();
+                this.inputLookup = new Dictionary<OutPoint, WalletIndexData>();
+
+                foreach (Wallet wallet in this.Wallets)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: You can use `SelectMany` here as well.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 1 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: It sort of implies it already - this method gets called both for mempool transactions and transactions found in a block (and you don't necessarily know which order these will occur or if it will even be called under both scenarios).

Therefore if `block` is non-`null` the transaction must be confirmed (unless there is a very inopportune reorg), and if it is `null` it probably isn't (there are potential race conditions with the mempool). So renaming the variable doesn't really add clarity I think, since the contents of the variable tells you everything you need to know.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 3 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Yes, the 4 scenarios determine which of the two possibilities (incoming or existing) are the 'better' one and acts accordingly. And yes, the index dictionary gets rebuilt as part of the removal process.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 6 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: I suppose so, although it might make it somewhat unclear what is being done. Basically the only reason I loop through everything is so that we obtain knowledge of the `Wallet` and `HdAccount` instances to use (as we already know the `HdAddress` and `TransactionData`, and they have no link to the other data to allow them to be efficiently looked up). With a change to the data returned by `scriptToAddressLookup` this would be completely unnecessary, but I didn't want to change too much more in this PR.

**Comment by maciejzaleski** [comment]
Regression tested this change and proven that it does solve the problem with unconfirmed transactions.
We should merge this to unblock testing of other areas.

**Comment by bokobza** [comment]
I really think this change should be blocked and more testing done on it.
The processing of wallet transactions is a very sensitive beast and this
change honestly doesn't look good at all and will probably make the wallet
at least 2x slower.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 8]
```diff
@@ -96,8 +96,12 @@ public class WalletManager : IWalletManager
         // we keep a couple of objects in memory:
         // 1. the list of unspent outputs for checking whether inputs from a transaction are being spent by our wallet and
         // 2. the list of addresses contained in our wallet for checking whether a transaction is being paid to the wallet.
+        // 3. a mapping of all transactions in the wallet, to allow rapid lookup by txid
+        // 4. a mapping of all inputs with their corresponding transactions, to facilitate rapid lookup
         private Dictionary<OutPoint, TransactionData> outpointLookup;
         internal ScriptToAddressLookup scriptToAddressLookup;
+        private Dictionary<uint256, WalletIndexData> txLookup;
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: This is not used, I'd say we should only add another index if we have a current use for it. also index by trx already exists in store via -txindex

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 9]
```diff
@@ -928,8 +934,9 @@ public void RemoveBlocks(ChainedHeader fork)
 
                 this.UpdateLastBlockSyncedHeight(fork);
 
-                // Reload the addresses and inputs lookup dictionaries.
+                // Reload the lookup dictionaries.
                 this.RefreshInputKeysLookupLock();
+                this.RefreshTxLookupLocked();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: This is a  bad idea, all index data should go in one lock or we will get race conditions.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 10]
```diff
@@ -1574,7 +1717,7 @@ public HashSet<(uint256, DateTimeOffset)> RemoveTransactionsByIdsLocked(string w
 
             lock (this.lockObject)
             {
-                IEnumerable<HdAccount> accounts = wallet.GetAccounts();
+                IEnumerable<HdAccount> accounts = wallet.GetAccounts(a => true);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: What does this do?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 11]
```diff
@@ -1003,6 +1010,50 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
 
             lock (this.lockObject)
             {
+                // Do a pre-scan of the incoming transaction's inputs to see if they're used in other wallet transactions already.
+                foreach (TxIn input in transaction.Inputs)
+                {
+                    // See if this input is being used by another wallet transaction present in the index.
+                    // The inputs themselves may not belong to the wallet, but the transaction data in the index has to be for a wallet transaction.
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Seems we only use the trx hash (uint256) I dont think we need to store this type `WalletIndexData` at all just reference the trx hash

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 12]
```diff
@@ -1003,6 +1010,50 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
 
             lock (this.lockObject)
             {
+                // Do a pre-scan of the incoming transaction's inputs to see if they're used in other wallet transactions already.
+                foreach (TxIn input in transaction.Inputs)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: `transaction.Inputs` is iterated twice can we not use the single loop?

Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: I wanted to avoid making any changes to the wallet data before it's determined whether or not it's an incoming double spend transaction. Note that the remainder of the method processes outputs first, which would modify the data, before proceeding onto the inputs.

I think enumerating the inputs once with a dictionary lookup per input (on average) is not too onerous.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 10 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: It filters out some of the accounts, in this case it will now return all of them. I think it was actually a bug that it wasn't being used previously, as it was leaving out the cold staking accounts.

I can't see a reason why you'd want to not search the cold staking accounts when removing by ID.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 9 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: These are within `lockObject` anyway right?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 11 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: I do see the benefit of indexing the whole structure though

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 9 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: ah right sorry my mistake yes correct.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 13]
```diff
@@ -1434,18 +1476,23 @@ public void LoadKeysLookupLock()
...
-                        // We only exclude from the list the confirmed spent UTXOs.
-                        foreach (TransactionData transaction in address.Transactions.Where(t => t.SpendingDetails?.BlockHeight == null))
+                        foreach (HdAddress address in account.GetCombinedAddresses())
                         {
-                            this.outpointLookup[new OutPoint(transaction.Id, transaction.Index)] = transaction;
+                            this.scriptToAddressLookup[address.ScriptPubKey] = address;
+                            if (address.Pubkey != null)
+                                this.scriptToAddressLookup[address.Pubkey] = address;
+
+                            foreach (TransactionData transaction in address.Transactions)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Why the change?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 14]
```diff
@@ -1434,18 +1476,23 @@ public void LoadKeysLookupLock()
             {
                 foreach (Wallet wallet in this.Wallets)
                 {
-                    IEnumerable<HdAddress> addresses = wallet.GetAllAddresses(a => true);
-                    foreach (HdAddress address in addresses)
+                    foreach (HdAccount account in wallet.GetAccounts(a => true))
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Why the split here? `GetAllAddresses` isn't working as expected?

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 15]
```diff
@@ -1105,6 +1134,12 @@ private void AddTransactionToWallet(Transaction transaction, TxOut utxo, int? bl
 
                 addressTransactions.Add(newTransaction);
                 this.AddInputKeysLookupLocked(newTransaction);
+
+                if (block == null)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Ok so here, if the transaction is by itself, you add the inputs in the lookup table against the newly created TransactionData...

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 16]
```diff
@@ -1132,8 +1167,15 @@ private void AddTransactionToWallet(Transaction transaction, TxOut utxo, int? bl
 
                 if (isPropagated)
                     foundTransaction.IsPropagated = true;
+
+                if (block != null)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: ...and here if the transaction is part of a block, you remove the corresponding TransactionData one from the index.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 17]
```diff
@@ -1003,6 +1006,32 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
...
+
+                            if (indexData.BlockHash != null)
+                            {
+                                // This should not happen as pre checks are done in mempool and consensus.
+                                throw new WalletException("The same inputs were found in two different confirmed transactions");
+                            }
+
+                            // This is a double spend we remove the unconfirmed trx
+                            this.RemoveTransactionsByIds(new[] { indexData.Id });
+                            this.inputLookup.Remove(input.PrevOut);
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Is the input not going to be removed from `RemoveTxLookupLocked` already?

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Not if it's a new TRX , I could move this code further down perhaps.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 13 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: There was new code here that I then removed later so yeah I can role this back actually.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 14 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: @zeptin claim this was a bug before, I think he is right, thoughts?

Comment (zeptin) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: The possible bug I was referring to is in `RemoveTransactionsByIdsLocked`; `a => true` was getting properly passed in `LoadKeysLookupLock` already, but it is missing there.

**Comment by dangershony** [comment]
This is how core deals with double spends

```
bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const uint256& block_hash, int posInBlock, bool fUpdate)
{
    const CTransaction& tx = *ptx;
    {
        AssertLockHeld(cs_wallet);

        if (!block_hash.IsNull()) {
            for (const CTxIn& txin : tx.vin) {
                std::pair<TxSpends::const_iterator, TxSpends::const_iterator> range = mapTxSpends.equal_range(txin.prevout);
                while (range.first != range.second) {
                    if (range.first->second != tx.GetHash()) {
                        WalletLogPrintf("Transaction %s (in block %s) conflicts with wallet transaction %s (both spend %s:%i)\n", tx.GetHash().ToString(), block_hash.ToString(), range.first->second.ToString(), range.first->first.hash.ToString(), range.first->first.n);
                        MarkConflicted(block_hash, range.first->second);
                    }
                    range.first++;
                }
            }
        }
```

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 18]
```diff
@@ -96,8 +96,10 @@ public class WalletManager : IWalletManager
         // we keep a couple of objects in memory:
         // 1. the list of unspent outputs for checking whether inputs from a transaction are being spent by our wallet and
         // 2. the list of addresses contained in our wallet for checking whether a transaction is being paid to the wallet.
+        // 3. a mapping of all inputs with their corresponding transactions, to facilitate rapid lookup
         private Dictionary<OutPoint, TransactionData> outpointLookup;
         internal ScriptToAddressLookup scriptToAddressLookup;
+        private Dictionary<OutPoint, TransactionData> inputLookup;
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: please add comment that it's supposed to be accessed under a lock `lockObject`

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: please add empty lines between the variables

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 13 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: If the change is unnecessary then please do

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 19]
```diff
@@ -1524,6 +1571,36 @@ public void RefreshInputKeysLookupLock()
             }
         }
 
+        /// <summary>
+        /// Add to the mapping of transactions kept in memory for faster lookups.
+        /// </summary>
+        private void AddTxLookupLocked(TransactionData transactionData, Transaction transaction)
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: If it has locked in the name it should have a comment in remarks section on which lock is required

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 18 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: please add comment on what this data structure is and what its used for

**Comment by noescape00** [comment]
Overall looks good but needs a bit of polishing
